nvs: avoid reading multiple times the same area

aurel32 opened the pull request:
The current NVS code checks for an empty or a closed ATE using the
_nvs_flash_cmp_const() function. This function loads the data and
compare them to a value. This means that when executed multiple on
the same area, the data get reloaded multiple time. This might have
a noticeable performance impact with an SPI flash.

Instead define two functions checking if an ATE is empty or closed
on an already read struct nvs_ate. Then replace the calls to
_nvs_flash_cmp_const() on struct nvs_ate by _nvs_flash_ate_rd() plus
those new functions. This also has the advantage of explicitly checking
for errors instead of testing the error and the result of the comparison
at the same time.

Tested on a Nucleo L432KC board. The maximum initialization time (ie
just before running the first garbage collector) goes down to 21ms from
32ms.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

Laczen reviewed (commented):
Hi @aurel32, why not one routine _nvs_ate_cmp_const(nvs_ate, value) that returns cmp values?

aurel32 commented:
@Laczen I have just pushed a new version changing the two functions by `_nvs_flash_cmp_const`.

Laczen commented:
@aurel32, I am working on a solution that would close the sector in the ate at sector end. This would contain the offset of the last ate in the sector. This should speed up the stepping through the fs.

aurel32 commented:
> @aurel32, I am working on a solution that would close the sector in the ate at sector end. This would contain the offset of the last ate in the sector. This should speed up the stepping through the fs.

Great if you can improve the speed that way. Even with that change, I think my patch would still improve the performances by reducing the number of reads. That's important for a SPI flash.

aurel32 committed:
nvs: avoid reading multiple times the same area

The current NVS code checks for an empty ATE using th
_nvs_flash_cmp_const() function. This function loads the data and
compare them to a value. This means that when executed multiple on the
same area, the data get reloaded multiple time. This might have a
noticeable performance impact with an SPI flash.

Instead define a function _nvs_ate_cmp_const to compare an already read
struct nvs_ate with a constant value. Then replace the calls to
_nvs_flash_cmp_const() on struct nvs_ate by _nvs_flash_ate_rd() followed
by _nvs_ate_cmp_const(). This also has the advantage of explicitly
checking for errors instead of testing the error and the result of the
comparison at the same time.

Tested on a Nucleo L432KC board with the nvs sample. The maximum
initialization time (ie just before running the first garbage collector)
goes down to 6213 µs from 7350 µs.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
@Laczen, Now that your NVS speed improvements have been merged, I have rebased my patch. I still see a 20% speed improvement when using the internal flash of a STM32L432KC SoC. I believe the improvement should be higher with an SPI flash, but I don't have a way to test that.

Laczen commented:
@aurel32, I wonder if the speed increase is the result of the different way the compare is done. Anyhow, I think the PR is an improvement for readability.

Laczen commented:
@aurel32, could you also add addr = 0 to the code so that #9767 can also be closed.

aurel32 committed:
nvs: workaround a GCC and Coverity warning

Following the recent NVS changes, the following warning now appear with
GCC 7.3 when building with -O2:

ZEPHYROOT/subsys/fs/nvs/nvs.c: In function 'nvs_reinit':
ZEPHYROOT/subsys/fs/nvs/nvs.c:92:36: warning: 'addr' may be used
uninitialized in this function [-Wmaybe-uninitialized]
  offset += fs->sector_size * (addr >> ADDR_SECT_SHIFT);
                              ~~~~~~^~~~~~~~~~~~~~~~~~~
ZEPHYROOT/subsys/fs/nvs/nvs.c:606:8: note: 'addr' was declared here
  u32_t addr;
        ^~~~

This was already reported by Coverity earlier as CID:187903.

In practice this can only happen if fs->sector_count equals 0, which is
not possible as checked in nvs_init(). At least in the GCC case, it
believes that k_mutex_lock(&fs->nvs_lock, K_FOREVER) could modify
fs->sector_count.

Workaround the issue by initializing addr to 0.

Fixes #9767

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>
